reference-tests/src/main/java/net/consensys/linea/ReferenceTestOutcomeRecorderTool.java

```diff
@@ -14,195 +14,200 @@
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Objects;
-import java.util.Set;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
 import lombok.Synchronized;
 import lombok.extern.slf4j.Slf4j;
 import net.consensys.linea.zktracer.json.JsonConverter;
 
 @Slf4j
 public class ReferenceTestOutcomeRecorderTool {
 
-  static JsonConverter jsonConverter = JsonConverter.builder().build();
-
-  @Synchronized
-  public static void mapAndStoreFailedReferenceTest(
-      String testName, List<String> logEventMessages, String jsonOutputFilename) {
-    Set<String> failedConstraints = getFailedConstraints(logEventMessages);
-    if (!failedConstraints.isEmpty()) {
-      readBlockchainReferenceTestsOutput(jsonOutputFilename)
-          .thenCompose(
-              jsonString -> {
-                JsonConverter jsonConverter = JsonConverter.builder().build();
-
-                BlockchainReferenceTestOutcome blockchainReferenceTestOutcome =
-                    getBlockchainReferenceTestOutcome(jsonString);
-
-                mapFailedConstraintsToTestsToModule(
-                    blockchainReferenceTestOutcome.modulesToConstraints(),
-                    failedConstraints,
-                    testName);
-
-                BlockchainReferenceTestOutcome finalBlockchainReferenceTestOutcome =
-                    new BlockchainReferenceTestOutcome(
-                        blockchainReferenceTestOutcome.failedCounter() + 1,
-                        blockchainReferenceTestOutcome.successCounter(),
-                        blockchainReferenceTestOutcome.modulesToConstraints());
-                jsonString = jsonConverter.toJson(finalBlockchainReferenceTestOutcome);
-                writeToJsonFile(jsonString, jsonOutputFilename);
-                return null;
-              });
+  public static final String JSON_INPUT_FILENAME = "failedBlockchainReferenceTests-input.json";
+  public static final String JSON_OUTPUT_FILENAME = "failedBlockchainReferenceTests.json";
+  public static JsonConverter jsonConverter = JsonConverter.builder().build();
+  public static volatile BlockchainReferenceTestOutcome testOutcomes =
+          new BlockchainReferenceTestOutcome(
+                  new AtomicInteger(0),
+                  new AtomicInteger(0),
+                  new AtomicInteger(0),
+                  new AtomicInteger(0),
+                  new ConcurrentHashMap<>());
```

Comment:
Mutating a record is an antipattern. It would be best to have individual counters and a concurrent map. At the end when you want to write test outcome create a BlockchainReferenceTestOutcome with values from counter and the map.